actors/builtin/payment_channel/payment_channel_actor.go

```diff
@@ -0,0 +1,247 @@
+	"github.com/filecoin-project/specs-actors/actors/builtin"
+	vmr "github.com/filecoin-project/specs-actors/actors/runtime"
+	indices "github.com/filecoin-project/specs-actors/actors/runtime/indices"
+	"github.com/filecoin-project/specs-actors/actors/serde"
+	"github.com/minio/blake2b-simd"
+)
+
+type PaymentChannelActor struct{}
+
+// type PaymentInfo struct {
+// 	PayChActor     address.Address
+// 	Payer          address.Address
+// 	ChannelMessage *cid.Cid
+
+// 	Vouchers []*types.SignedVoucher
+// }
+
+/////////////////
+// Constructor //
+/////////////////
+
+type PCAConstructorParams struct {
+	To addr.Address
+}
+
+func (pca *PaymentChannelActor) Constructor(rt vmr.Runtime, params *PCAConstructorParams) *vmr.EmptyReturn {
+
+	var st PaymentChannelActorState
+	rt.State().Transaction(&st, func() interface{} {
+		st.From = rt.ImmediateCaller()
+		st.To = params.To
+		st.LaneStates = make(map[int64]*LaneState)
+		return nil
+	})
+	return &vmr.EmptyReturn{}
+}
+
+////////////////////////////////////////////////////////////////////////////////
+// Payment Channel state operations
+////////////////////////////////////////////////////////////////////////////////
+
+type PCAUpdateChannelStateParams struct {
+	Sv     SignedVoucher
+	Secret []byte
+	Proof  []byte
+}
+
+func hash(b []byte) []byte {
+	s := blake2b.Sum256(b)
+	return s[:]
+}
+
+type PaymentVerifyParams struct {
+	Extra []byte
+	Proof []byte
+}
+
+func (pca *PaymentChannelActor) UpdateChannelState(rt vmr.Runtime, params *PCAUpdateChannelStateParams) *vmr.EmptyReturn {
+
+	var st PaymentChannelActorState
+	rt.State().Transaction(&st, func() interface{} {
+
+		sv := params.Sv
+
+		vb, nerr := sv.SigningBytes()
+		if nerr != nil {
+			rt.Abort(1, "failed to serialize signedvoucher")
+		}
+
+		if !rt.Syscalls().VerifySignature(*sv.Signature, st.From, vb) {
+			rt.Abort(2, "voucher signature invalid")
+		}
+
+		if rt.CurrEpoch() < sv.TimeLock {
+			rt.Abort(3, "cannot use this voucher yet!")
+		}
+
+		if len(sv.SecretPreimage) > 0 {
+			if !bytes.Equal(hash(params.Secret), sv.SecretPreimage) {
+				rt.Abort(4, "incorrect secret!")
+			}
+		}
+
+		if sv.Extra != nil {
+
+			_, code := rt.Send(
+				sv.Extra.Actor,
+				sv.Extra.Method,
+				serde.MustSerializeParams(
+					&PaymentVerifyParams{
+						sv.Extra.Data,
+						params.Proof,
+					},
+				),
+				abi.NewTokenAmount(0),
+			)
+			vmr.RequireSuccess(rt, code, "spend voucher verification failed")
+		}
+
+		ls, ok := st.LaneStates[sv.Lane]
+		// create voucher lane if it does not already exist
+		if !ok {
+			ls = new(LaneState)
+			ls.Redeemed = big.NewInt(0)
+			st.LaneStates[sv.Lane] = ls
+		}
+		if ls.Closed {
+			rt.Abort(5, "cannot redeem a voucher on a closed lane")
+		}
+
+		if ls.Nonce > sv.Nonce {
+			rt.Abort(6, "voucher has an outdated nonce, cannot redeem")
+		}
+
+		// The next section actually calculates the payment amounts to update the payment channel state
```

Comment:
The below is a bit tricky and deserves review @anorth

Reply:
I can't follow this giant method, but it's not your job to refactor it. I trust Lotus have tests we'll be able to port into here.